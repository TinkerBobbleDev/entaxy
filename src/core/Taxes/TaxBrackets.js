export const TaxBrackets = {
	2017: {
		federal: [
			{amountUpTo: 11635.00, tax: 0},
			{amountUpTo: 46605.00, tax: 15.00},
			{amountUpTo: 93208.00, tax: 20.50},
			{amountUpTo: 144489.00, tax: 26.00},
			{amountUpTo: 205842.00, tax: 29.00},
			{amountUpTo: 10000000, tax: 33.00}
		],
		'Ontario': [
			{amountUpTo: 10171.00, tax: 0},
			{amountUpTo: 42960.00, tax: 5.05},
			{amountUpTo: 83923.00, tax: 9.15},
			{amountUpTo: 150000.00, tax: 11.16},
			{amountUpTo: 220000.00, tax: 12.16},
			{amountUpTo: 10000000, tax: 13.16}
		],
	}
}

// Return an array with income and corresponding tax
export const IncomeTaxData = ({year, province, income}) => {
	income = Number.parseFloat(income)
	const min = 0
	const max = income > 0 ? income + 25000 : 250000
	const step = 100

	let data = []
	let bracketIndex = 0
	for (var currIncome = min; currIncome < max; currIncome += step) {
		// Add the exact bracket ammount
		if(currIncome >= TaxBrackets[year][province][bracketIndex].amountUpTo) {
			data.push({
				income: TaxBrackets[year][province][bracketIndex].amountUpTo,
				tax: calculateTotalTax(year, province, TaxBrackets[year][province][bracketIndex].amountUpTo)
			})
			bracketIndex++;
		}

		data.push({
			income: currIncome,
			tax: calculateTotalTax(year, province, currIncome)
		})
	}
	return data
}

export const taxBracketData = (year, province) => {
	const data = []

	for(var b = 0; b < TaxBrackets[year][province].length-1; b++) {
		data.push({
			type: 'provincial',
			income: TaxBrackets[year][province][b].amountUpTo,
			tax: TaxBrackets[year][province][b+1].tax
		})
	}
	for(b = 0; b < TaxBrackets[year]['federal'].length-1; b++) {
		data.push({
			type: 'federal',
			income: TaxBrackets[year]['federal'][b].amountUpTo,
			tax: TaxBrackets[year]['federal'][b+1].tax
		})
	}

	data.sort((a, b) => {return a.income - b.income})

	var tax = {provincial: 0, federal: 0}
	for(var i = 0; i < data.length; i++) {
		switch(data[i].type) {
			case 'provincial':
				tax.provincial = data[i].tax
				break
			case 'federal':
				tax.federal = data[i].tax
				break
			default:
		}
		data[i].tax = tax.federal + tax.provincial
	}

	return data
}

export const calculateTotalTax = (year, province, income) => {
	return calculateTax(year, 'federal', income) + calculateTax(year, province, income)
}

export const calculateTax = (year, province, income) => {
	let tax = 0

	for(var b = 0; b < TaxBrackets[year][province].length; b++) {
		var bracket = TaxBrackets[year][province][b]

		if(bracket.tax === 0) {
			if(income < bracket.amountUpTo) {
				break
			}
		} else {
			var prevBracketAmount = TaxBrackets[year][province][b-1].amountUpTo
			if(bracket.amountUpTo === null) {
				// Pay the higher tax for the rest of the incomde
				tax += (income - prevBracketAmount) * (bracket.tax / 100)
			} else if(income > bracket.amountUpTo) {
				// Pay the full amount for this bracket
				tax += (bracket.amountUpTo - prevBracketAmount) * (bracket.tax / 100)
			} else {
				// Pay only the difference
				tax += (income - prevBracketAmount) * (bracket.tax / 100)
				break
			}
		}
	}
	return tax
}

// The marginal tax rate is the amount of tax paid
// on any additional dollar made, up to the next tax bracket.
export const marginalTax = (year, province, income) => {
	for(var b = 0; b < TaxBrackets[year][province].length; b++) {
		if(income < TaxBrackets[year][province][b].amountUpTo) {
			return TaxBrackets[year][province][b].tax / 100
		}
	}
}

export const totalMarginalTax = (year, province, income) => {
	return marginalTax(year, 'federal', income) + marginalTax(year, province, income)
}
