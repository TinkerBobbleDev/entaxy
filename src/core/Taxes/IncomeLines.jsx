import React from 'react'
import PropTypes from 'prop-types'
import { Line } from '@vx/shape'
import { Point } from '@vx/point'
import { Motion, spring } from 'react-motion'
import teal from '@material-ui/core/colors/teal'
import { TaxBrackets, calculateTotalTax } from './lib/TaxBrackets'

const IncomeLines = ({
  income,
  country,
  year,
  region,
  xScale,
  yScale,
  margin,
  height
}) => {
  const yMax = height - margin.top - margin.bottom
  const left = xScale(income)
  const top = yScale(calculateTotalTax(TaxBrackets[country][year], region, income))

  return (
    <Motion
      defaultStyle={{ left, top }}
      style={{
        left: spring(left),
        top: spring(top)
      }}
    >
      {style => (
        <g key="IncomeLines">
          <Line
            key="IncomeLine-vertical"
            from={new Point({ x: style.left, y: yMax })}
            to={new Point({ x: style.left, y: style.top })}
            stroke={teal[500]}
            strokeWidth={1}
          />
          <Line
            key="IncomeLine-horizontal"
            from={new Point({ x: 0, y: style.top })}
            to={new Point({ x: style.left, y: style.top })}
            stroke={teal[500]}
            strokeWidth={1}
          />
          <circle
            cx={style.left - 1}
            cy={style.top + 1}
            r={4}
            fill="white"
            stroke={teal[500]}
            strokeWidth="1.5"
            fillOpacity={1}
            strokeOpacity={1}
          />
        </g>
      )}
    </Motion>
  )
}

IncomeLines.propTypes = {
  income: PropTypes.number.isRequired,
  country: PropTypes.string.isRequired,
  year: PropTypes.number.isRequired,
  region: PropTypes.string,
  xScale: PropTypes.func.isRequired,
  yScale: PropTypes.func.isRequired,
  margin: PropTypes.object.isRequired,
  height: PropTypes.number.isRequired
}

IncomeLines.defaultProps = {
  region: null
}

export default IncomeLines
