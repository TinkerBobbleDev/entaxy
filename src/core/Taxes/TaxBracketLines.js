import React from 'react'
import { Line } from '@vx/shape';
import { Point } from '@vx/point';
import { taxBracketData } from './TaxBrackets'
import red from 'material-ui/colors/red';
import blueGrey from 'material-ui/colors/blueGrey';

const TaxBracketLines = (
  year,
  province,
  xScale,
  yScale,
  margin,
  width,
  height
) => {
  const rows = []
  const yMax = height - margin.top - margin.bottom
  const xMax = width - margin.left - margin.right
  const data = taxBracketData(year, province)

  for(var d=0; d < data.length; d++){
    var left = xScale(data[d].income)
    var right = data[d+1] ? xScale(data[d+1].income) : xMax

    rows.push(
      <Line
        key={"bracket-" + d}
        from={new Point({x: left, y: yMax})}
        to={new Point({x: left, y: 10})}
        stroke={red[100]}
        strokeWidth={1}
      />
    )
    rows.push(
      <text
        fill={blueGrey[800]}
        textAnchor="middle"
        x={left + (right - left)/2}
        y={0}
        dy=".33em"
        fontSize={9}
        fontFamily={'Roboto'}
      >
      {data[d].tax}%
    </text>

    )
  }
  return  <g key={"TaxBracketLines"}>{rows}</g>
}

export default TaxBracketLines
