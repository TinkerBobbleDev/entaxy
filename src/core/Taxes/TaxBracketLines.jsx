import React from 'react'
import PropTypes from 'prop-types'
import { Line } from '@vx/shape'
import { Point } from '@vx/point'
import red from 'material-ui/colors/red'
import blueGrey from 'material-ui/colors/blueGrey'
import { TaxBrackets, taxBracketData } from './lib/TaxBrackets'

const TaxBracketLines = ({
  country,
  year,
  region,
  xScale,
  margin,
  width,
  height
}) => {
  const rows = []
  const yMax = height - margin.top - margin.bottom
  const xMax = width - margin.left - margin.right
  const data = taxBracketData(TaxBrackets[country][year], region)

  for (let d = 0; d < data.length; d += 1) {
    const left = xScale(data[d].income)
    const right = data[d + 1] ? xScale(data[d + 1].income) : xMax

    const line = (
      <Line
        key={`bracket-line${d}`}
        from={new Point({ x: left, y: yMax })}
        to={new Point({ x: left, y: 10 })}
        stroke={red[100]}
        strokeWidth={1}
      />
    )

    const text = (
      <text
        key={`bracket-text${d}`}
        fill={blueGrey[800]}
        textAnchor="middle"
        x={left + ((right - left) / 2)}
        y={0}
        dy=".33em"
        fontSize={9}
        fontFamily="Roboto"
      >
        {data[d].tax}
      </text>
    )

    rows.push(line)
    rows.push(text)
  }
  return <g key="TaxBracketLines">{rows}</g>
}

TaxBracketLines.propTypes = {
  country: PropTypes.string.isRequired,
  year: PropTypes.number.isRequired,
  region: PropTypes.string,
  xScale: PropTypes.func.isRequired,
  margin: PropTypes.object.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired
}

TaxBracketLines.defaultProps = {
  region: null
}

export default TaxBracketLines
