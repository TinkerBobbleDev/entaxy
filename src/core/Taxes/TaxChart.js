import React from 'react'
import { extent, max, bisector } from 'd3-array';
import { Group } from '@vx/group';
import { scaleLinear } from '@vx/scale';
import { LinePath, Bar } from '@vx/shape';
import { AxisLeft, AxisBottom } from '@vx/axis';
import { curveBasis } from '@vx/curve';
import { withTooltip } from '@vx/tooltip'
import { localPoint } from '@vx/event';
import { Line } from '@vx/shape';
import { Point } from '@vx/point';

import IncomeLines from './IncomeLines'
import TaxCreditLines from './TaxCreditLines'
import TaxBracketLines from './TaxBracketLines'
import { IncomeTaxData } from './TaxBrackets'
import TaxTooltips from './TaxTooltips'

import blueGrey from 'material-ui/colors/blueGrey';

const colors = ['rgb(107, 157, 255)', 'rgb(252, 137, 159)']

const province = 'Ontario'

// Bounds
const margin = {
  top: 20,
  bottom: 40,
  left: 70,
  right: 0,
};

const TaxChart = ({
  width,
  height,
  year,
  income,
  rrsp,
  tooltipOpen,
  tooltipLeft,
  tooltipTop,
  tooltipData,
  showTooltip,
  hideTooltip,
}) => {
  const data = IncomeTaxData({year: year, province: province, income: income})
  const xMax = width - margin.left - margin.right
  const yMax = height - margin.top - margin.bottom

  const x = d => {
    if (d === undefined) {
      return 0
    } else {
      return d.income;
    }
  }
  const y = d => {
    if (d === undefined) {
      return 0
    } else {
      return d.tax;
    }
  }

  const xScale = scaleLinear({
    range: [0, xMax],
    domain: extent(data, x)
  })

  const yScale = scaleLinear({
    range: [yMax, 0],
    domain: [0, max(data, y)],
  })

  return (
    <div>
      <svg width={width} height={height} ref={s => (this.svg = s)}>
        <Group top={margin.top} left={margin.left}>
          {TaxBracketLines(year, province, xScale, yScale, margin, width, height)}
          <LinePath
            data={data}
            xScale={xScale}
            yScale={yScale}
            x={x}
            y={y}
            stroke={blueGrey[100]}
            strokeWidth={4}
            curve={curveBasis}
          />
          <AxisLeft
            scale={yScale}
            top={0}
            left={0}
            label={'Tax'}
            labelOffset={50}
            labelProps={{fontFamily: 'Roboto', fontSize: 12 }}
            stroke={'#dddddd'}
            tickLabelProps={(val, i) => ({
              x: -15,
              textAnchor: 'end',
              fontFamily: 'Roboto',
              fontSize: 10
            })}
            tickFormat={yScale.tickFormat(10, "$,f")}
            tickStroke={'#dddddd'}
          />
          <AxisBottom
            scale={xScale}
            top={yMax}
            label={'Income'}
            labelProps={{fontFamily: 'Roboto', fontSize: 12 }}
            stroke={'#dddddd'}
            tickLabelProps={(val, i) => ({
              textAnchor: 'middle',
              fontFamily: 'Roboto',
              fontSize: 10
            })}
            tickFormat={xScale.tickFormat(10, "$,f")}
            tickStroke={'#dddddd'}
          />
          {TaxCreditLines(income, rrsp, year, province, xScale, yScale, margin, width, height)}
          {IncomeLines(income, year, province, xScale, yScale, margin, width, height)}
          <Bar
            data={data}
            width={width}
            height={height - margin.bottom}
            fill="transparent"
            onMouseLeave={data => event => hideTooltip()}
            onMouseMove={data => event => {
              // Figure out he data index from the mouse hover
              const { x: xPoint } = localPoint(this.svg, event);
              const x0 = xScale.invert(xPoint - margin.left) ;
              const index = bisector(d => x(d)).left(data, x0);
              showTooltip({
                tooltipData: data[index],
                tooltipLeft: xScale(x(data[index])) + margin.left,
                tooltipTop: yScale(y(data[index])) + margin.top
              });
            }}
          />
        </Group>
        {tooltipData &&
          <g key={"g-"}>
            <Line
              key={"vertical-"}
              from={new Point({x: tooltipLeft, y: yMax + margin.bottom})}
              to={new Point({x: tooltipLeft, y: margin.top})}
              stroke={"rgb(220, 220, 220)"}
            />
            <Line
              key={"horizontal-"}
              from={new Point({x: margin.left, y: tooltipTop})}
              to={new Point({x: xMax + margin.left, y: tooltipTop})}
              stroke={"rgb(220, 220, 220)"}
            />
            <circle
              cx={tooltipLeft}
              cy={tooltipTop}
              r={12}
              fill={colors[0]}
              stroke={colors[0]}
              strokeWidth=".6"
              fillOpacity={1 / 12}
              strokeOpacity={1 / 2}
            />
            {/* Inner circle */}
            <circle
              cx={tooltipLeft}
              cy={tooltipTop}
              r={4}
              fill="white"
              stroke={colors[0]}
              strokeWidth="1.5"
              fillOpacity={1}
              strokeOpacity={1}
            />
          </g>
        }
      </svg>
      {tooltipData &&
        <TaxTooltips
          tooltipOpen={tooltipOpen}
          data={tooltipData}
          top={tooltipTop}
          left={tooltipLeft}
          margin={margin}
        />}
    </div>
  )
}

export default withTooltip(TaxChart)
