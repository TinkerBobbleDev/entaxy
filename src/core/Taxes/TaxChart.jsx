import React from 'react'
import PropTypes from 'prop-types'
import { withParentSize } from '@vx/responsive'
import { extent, max } from 'd3-array'
import { Group } from '@vx/group'
import { scaleLinear } from '@vx/scale'
import { LinePath } from '@vx/shape'
import { AxisBottom, AxisLeft } from '@vx/axis'
import { curveBasis } from '@vx/curve'
import blueGrey from '@material-ui/core/colors/blueGrey'

import IncomeLines from './IncomeLines'
import TaxCreditLines from './TaxCreditLines'
import TaxBracketLines from './TaxBracketLines'
import { TaxBrackets, incomeTaxData } from './lib/TaxBrackets'
import TaxTooltips from './TaxTooltips'

// Bounds
const margin = {
  top: 20,
  bottom: 40,
  left: 70,
  right: 0
}

const x = d => d.income
const y = d => d.tax

const TaxChart = ({
  parentWidth,
  parentHeight,
  country,
  year,
  region,
  income,
  rrsp,
  taxBeforeCredits
}) => {
  const width = parentWidth
  const height = parentHeight - margin.bottom
  const xMax = width - margin.left - margin.right
  const yMax = height - margin.top - margin.bottom
  const data = incomeTaxData(TaxBrackets[country][year], region, income)
  const xScale = scaleLinear({ range: [0, xMax], domain: extent(data, x) })
  const yScale = scaleLinear({ range: [yMax, 0], domain: [0, max(data, y)] })

  return (
    <div>
      <svg width={width} height={height}>
        <Group top={margin.top} left={margin.left}>
          <TaxBracketLines
            country={country}
            year={year}
            region={region}
            xScale={xScale}
            yScale={yScale}
            margin={margin}
            width={width}
            height={height}
          />
          <LinePath
            data={data}
            xScale={xScale}
            yScale={yScale}
            x={x}
            y={y}
            stroke={blueGrey[100]}
            strokeWidth={4}
            curve={curveBasis}
          />
          <AxisLeft
            scale={yScale}
            top={0}
            left={0}
            label="Tax"
            labelOffset={50}
            labelProps={{ fontFamily: 'Roboto', fontSize: 12 }}
            stroke="#dddddd"
            tickLabelProps={() => ({
              x: -15,
              textAnchor: 'end',
              fontFamily: 'Roboto',
              fontSize: 10
            })}
            tickFormat={yScale.tickFormat(10, '$,f')}
            tickStroke="#dddddd"
          />
          <AxisBottom
            scale={xScale}
            top={yMax}
            label="Income"
            labelProps={{ fontFamily: 'Roboto', fontSize: 12 }}
            stroke="#dddddd"
            tickLabelProps={() => ({
              textAnchor: 'middle',
              fontFamily: 'Roboto',
              fontSize: 10
            })}
            tickFormat={xScale.tickFormat(10, '$,f')}
            tickStroke="#dddddd"
          />
          <TaxCreditLines
            income={income}
            credits={rrsp}
            country={country}
            year={year}
            region={region}
            xScale={xScale}
            yScale={yScale}
            margin={margin}
            width={width}
            height={height}
          />
          <IncomeLines
            income={income}
            country={country}
            year={year}
            region={region}
            xScale={xScale}
            yScale={yScale}
            margin={margin}
            width={width}
            height={height}
          />
        </Group>
      </svg>
      <TaxTooltips
        data={{ income, tax: taxBeforeCredits }}
        top={yScale(taxBeforeCredits) + 100}
        left={Math.min(xMax - 100, xScale(income) + margin.left)}
      />
    </div>
  )
}

TaxChart.propTypes = {
  parentWidth: PropTypes.number.isRequired,
  parentHeight: PropTypes.number.isRequired,
  country: PropTypes.string.isRequired,
  year: PropTypes.number.isRequired,
  region: PropTypes.string,
  income: PropTypes.number.isRequired,
  rrsp: PropTypes.number.isRequired,
  taxBeforeCredits: PropTypes.number.isRequired
}

TaxChart.defaultProps = {
  region: null
}

export default withParentSize(TaxChart)
