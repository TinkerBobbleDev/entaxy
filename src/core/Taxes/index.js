import React from 'react'
import { withStyles } from 'material-ui/styles';
import Grid from 'material-ui/Grid'
import Paper from 'material-ui/Paper'
import Input, { InputLabel } from 'material-ui/Input';
import { FormControl } from 'material-ui/Form';
import { ParentSize } from "@vx/responsive";
import TaxChart from './TaxChart'
import CurrencyFormat from '../../common/CurrencyFormat'
import { calculateTotalTax, totalMarginalTax } from './TaxBrackets'
import Typography from 'material-ui/Typography';

const styles = theme => ({
  root: {
    flexGrow: 1,
  },
  paper: {
    margin: '10px 5px',
    padding: '20px',
  },
})

class Taxes extends React.Component {

  constructor(props){
    super(props);
	  const income = 90000
	  const rrsp = 10000
	  const taxable_income = income - rrsp
	  const tax_amount = calculateTotalTax(2017, 'Ontario', taxable_income)

    this.state = {
      province: 'Ontario',
      income: income,
      rrsp: rrsp,
      taxable_income:  taxable_income,
      tax_amount: tax_amount,
      marginal_tax_rate: totalMarginalTax(2017, 'Ontario', taxable_income),
      average_tax_rate: tax_amount / (taxable_income)
    }
  }




  handleChange = name => event => {
    let income = name === 'income' ? event.target.value : this.state.income
    let rrsp = name === 'rrsp' ? event.target.value : this.state.rrsp
    let taxable_income = income - rrsp
    let tax_amount = calculateTotalTax(2017, this.state.province, taxable_income)
    this.setState({
      [name]: event.target.value,
      taxable_income: taxable_income,
      tax_amount: tax_amount,
      marginal_tax_rate: totalMarginalTax(2017, this.state.province, taxable_income),
      average_tax_rate: tax_amount / (income - rrsp)
    })
  }


  render() {
    const { classes } = this.props
    var currencyFormatter = new Intl.NumberFormat('en-CA', { style: 'currency', currency: 'CAD' })
    var percentFormatter = new Intl.NumberFormat('en-CA', {style: 'percent'})
    return (
      <div className={classes.root}>
        <Grid container spacing={0}>
          <Grid item xs={10}>
            <Paper className={classes.paper}>
              <Typography variant="headline" gutterBottom align="center">
                Taxes for 2017
              </Typography>
              <ParentSize>
                {parent => (
                  <TaxChart
                    year={2017}
                    width={parent.width}
                    height={500}
                    income={this.state.income}
                    rrsp={this.state.rrsp}
                  />)
                }
              </ParentSize>
            </Paper>
          </Grid>
          <Grid item xs={2}>
            <Paper className={classes.paper}>
              <FormControl fullWidth className={classes.formControl}>
                <InputLabel htmlFor="income">Your income</InputLabel>
                <Input
                  value={this.state.income}
                  onChange={this.handleChange('income')}
                  inputComponent={CurrencyFormat}
                  className={classes.input}
                  inputProps={{
                    'aria-label': 'Income for 2017',
                  }}
                />
              </FormControl>
            </Paper>
            <Paper className={classes.paper}>
              <Typography variant="subheading">
                Tax credits
              </Typography>
              <FormControl fullWidth className={classes.formControl}>
                <InputLabel htmlFor="rrsp">RRSP contribution</InputLabel>
                <Input
                  value={this.state.rrsp}
                  onChange={this.handleChange('rrsp')}
                  inputComponent={CurrencyFormat}
                  className={classes.rrsp}
                  inputProps={{
                    'aria-label': 'RRSP for 2017',
                  }}
                />
              </FormControl>
            </Paper>
            <Paper className={classes.paper}>
              <Typography>
                <strong>Taxable income:</strong> {currencyFormatter.format(this.state.income - this.state.rrsp)}
              </Typography>
              <Typography>
                <strong>Tax amount:</strong> {currencyFormatter.format(this.state.tax_amount)}
              </Typography>
              <Typography>
                <strong>Marginal Tax Rate:</strong> {percentFormatter.format(this.state.marginal_tax_rate)}
              </Typography>
              <Typography>
                <strong>Average Tax Rate:</strong> {percentFormatter.format(this.state.average_tax_rate)}
              </Typography>
            </Paper>
          </Grid>
        </Grid>
      </div>
      )
    }
}

export default withStyles(styles)(Taxes);
