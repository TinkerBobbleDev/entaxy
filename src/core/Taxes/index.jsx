import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from 'material-ui/styles'
import Grid from 'material-ui/Grid'
import Paper from 'material-ui/Paper'
import Input, { InputLabel } from 'material-ui/Input'
import { FormControl } from 'material-ui/Form'
import Typography from 'material-ui/Typography'
import TaxChart from './TaxChart'
import CurrencyFormat from '../../common/CurrencyFormat/index'
import { TaxBrackets, calculateTotalTax, totalMarginalTax } from './lib/TaxBrackets'

const styles = () => ({
  root: {
    flexGrow: 1
  },
  paper: {
    margin: '10px 5px',
    padding: '20px'
  },
  chart: {
    height: '500px'
  }
})

class Taxes extends React.Component {
  constructor(props) {
    super(props)
    const country = 'Canada'
    const year = 2017
    const region = 'Ontario'
    const income = 90000
    const rrsp = 10000
    const taxableIncome = income - rrsp
    const taxAmount = calculateTotalTax(TaxBrackets[country][year], region, taxableIncome)
    const taxBeforeCredits = calculateTotalTax(TaxBrackets[country][year], region, income)

    this.state = {
      country,
      year,
      region,
      income,
      rrsp,
      taxableIncome,
      taxBeforeCredits,
      taxAmount,
      marginalTaxRate: totalMarginalTax(TaxBrackets[country][year], region, taxableIncome),
      averageTaxRate: taxAmount / (taxableIncome)
    }
  }

  handleChange = name => (event) => {
    let income = name === 'income' ? parseFloat(event.target.value) : this.state.income
    if (Number.isNaN(income)) { income = 0 }
    let rrsp = name === 'rrsp' ? parseFloat(event.target.value) : this.state.rrsp
    if (Number.isNaN(rrsp)) { rrsp = 0 }
    if (rrsp > income) { rrsp = income }
    const taxableIncome = income - rrsp
    const { country, year, region } = this.state
    const taxAmount = calculateTotalTax(TaxBrackets[country][year], region, taxableIncome)
    const taxBeforeCredits = calculateTotalTax(TaxBrackets[country][year], region, income)
    this.setState({
      income,
      rrsp,
      taxableIncome,
      taxAmount,
      taxBeforeCredits,
      marginalTaxRate: totalMarginalTax(TaxBrackets[country][year], region, taxableIncome),
      averageTaxRate: taxableIncome > 0 ? taxAmount / taxableIncome : 0
    })
  }

  render() {
    const { classes } = this.props
    const {
      country,
      year,
      region,
      income,
      rrsp,
      taxAmount,
      taxableIncome,
      taxBeforeCredits,
      marginalTaxRate,
      averageTaxRate
    } = this.state
    const currencyFormatter = new Intl.NumberFormat('en-CA', { style: 'currency', currency: 'CAD' })
    const percentFormatter = new Intl.NumberFormat('en-CA', { style: 'percent', minimumFractionDigits: 2 })
    return (
      <div className={classes.root}>
        <Grid container spacing={0}>
          <Grid item xs={10}>
            <Paper className={`${classes.paper} ${classes.chart}`}>
              <Typography variant="headline" gutterBottom align="center">
                Taxes for {region} {year}
              </Typography>
              <TaxChart
                country={country}
                year={year}
                region={region}
                income={income}
                rrsp={rrsp}
                taxBeforeCredits={taxBeforeCredits}
              />
            </Paper>
          </Grid>
          <Grid item xs={2}>
            <Paper className={classes.paper}>
              <FormControl fullWidth className={classes.formControl}>
                <InputLabel htmlFor="income">Your income</InputLabel>
                <Input
                  value={income}
                  onChange={this.handleChange('income')}
                  inputComponent={CurrencyFormat}
                  className={classes.input}
                  inputProps={{
                    'aria-label': `Your income for ${year}`
                  }}
                />
              </FormControl>
            </Paper>
            <Paper className={classes.paper}>
              <Typography variant="subheading">
                Tax credits
              </Typography>
              <FormControl fullWidth className={classes.formControl}>
                <InputLabel htmlFor="rrsp">RRSP contribution</InputLabel>
                <Input
                  value={rrsp}
                  onChange={this.handleChange('rrsp')}
                  inputComponent={CurrencyFormat}
                  className={classes.rrsp}
                  inputProps={{
                    'aria-label': `RRSP for ${year}`
                  }}
                />
              </FormControl>
            </Paper>
            <Paper className={classes.paper}>
              <Typography>
                <strong>Taxable income:</strong> {currencyFormatter.format(taxableIncome)}
              </Typography>
              <Typography>
                <strong>Tax amount:</strong> {currencyFormatter.format(taxAmount)}
              </Typography>
              <Typography>
                <strong>Marginal Tax Rate:</strong> {percentFormatter.format(marginalTaxRate)}
              </Typography>
              <Typography>
                <strong>Average Tax Rate:</strong> {percentFormatter.format(averageTaxRate)}
              </Typography>
            </Paper>
          </Grid>
        </Grid>
      </div>
    )
  }
}

Taxes.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(Taxes)
