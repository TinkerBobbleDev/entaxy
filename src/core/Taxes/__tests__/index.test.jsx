import React from 'react'
import renderer from 'react-test-renderer'
import Taxes from '../'

describe('Taxes', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<Taxes />)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
