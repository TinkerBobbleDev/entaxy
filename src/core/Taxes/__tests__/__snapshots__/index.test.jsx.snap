// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Taxes matches snapshot 1`] = `
"<div className={[undefined]}>
  <WithStyles(Header) />
  <WithStyles(Grid) container={true} spacing={0}>
    <WithStyles(Grid) item={true} xs={10}>
      <WithStyles(Paper) className=\\"undefined undefined\\">
        <WithStyles(Typography) variant=\\"headline\\" gutterBottom={true} align=\\"center\\">
          Taxes for 
          Ontario
           
          2017
        </WithStyles(Typography)>
        <WrappedComponent country=\\"Canada\\" year={2017} region=\\"Ontario\\" income={90000} rrsp={10000} taxBeforeCredits={20282.9731} windowResizeDebounceTime={300} />
      </WithStyles(Paper)>
    </WithStyles(Grid)>
    <WithStyles(Grid) item={true} xs={2}>
      <WithStyles(Paper) className={[undefined]}>
        <WithStyles(FormControl) fullWidth={true} className={[undefined]}>
          <WithStyles(InputLabel) htmlFor=\\"income\\">
            Your income
          </WithStyles(InputLabel)>
          <WithStyles(Input) value={90000} onChange={[Function]} inputComponent={[Function: CurrencyFormat]} className={[undefined]} inputProps={{...}} />
        </WithStyles(FormControl)>
      </WithStyles(Paper)>
      <WithStyles(Paper) className={[undefined]}>
        <WithStyles(Typography) variant=\\"subheading\\">
          Tax credits
        </WithStyles(Typography)>
        <WithStyles(FormControl) fullWidth={true} className={[undefined]}>
          <WithStyles(InputLabel) htmlFor=\\"rrsp\\">
            RRSP contribution
          </WithStyles(InputLabel)>
          <WithStyles(Input) name=\\"rrsp\\" value={10000} onChange={[Function]} inputComponent={[Function: CurrencyFormat]} className={{...}} inputProps={{...}} />
        </WithStyles(FormControl)>
      </WithStyles(Paper)>
      <WithStyles(Paper) className={[undefined]}>
        <WithStyles(Typography)>
          <strong>
            Taxable income:
          </strong>
           
          CA$80,000.00
        </WithStyles(Typography)>
        <WithStyles(Typography)>
          <strong>
            Tax amount:
          </strong>
           
          CA$17,205.49
        </WithStyles(Typography)>
        <WithStyles(Typography)>
          <strong>
            Marginal Tax Rate:
          </strong>
           
          29.65%
        </WithStyles(Typography)>
        <WithStyles(Typography)>
          <strong>
            Average Tax Rate:
          </strong>
           
          21.51%
        </WithStyles(Typography)>
      </WithStyles(Paper)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</div>"
`;
